crypto/ml_kem/ml_kem.c

```diff
@@ -0,0 +1,1732 @@
+
+#  define ML_KEM_512_U_VECTOR_BYTES \
+    ML_KEM_U_VECTOR_BYTES(ML_KEM_512_RANK, ML_KEM_512_DU)
+#  define ML_KEM_768_U_VECTOR_BYTES \
+    ML_KEM_U_VECTOR_BYTES(ML_KEM_768_RANK, ML_KEM_768_DU)
+#  define ML_KEM_1024_U_VECTOR_BYTES \
+    ML_KEM_U_VECTOR_BYTES(ML_KEM_1024_RANK, ML_KEM_1024_DU)
+
+#  define ML_KEM_512_V_SCALAR_BYTES \
+    ML_KEM_V_SCALAR_BYTES(ML_KEM_512_DV)
+#  define ML_KEM_768_V_SCALAR_BYTES \
+    ML_KEM_V_SCALAR_BYTES(ML_KEM_768_DV)
+#  define ML_KEM_1024_V_SCALAR_BYTES \
+    ML_KEM_V_SCALAR_BYTES(ML_KEM_1024_DV)
+
+#  define ML_KEM_512_CIPHERTEXT_BYTES \
+    (ML_KEM_512_U_VECTOR_BYTES + ML_KEM_512_V_SCALAR_BYTES)
+#  define ML_KEM_768_CIPHERTEXT_BYTES \
+    (ML_KEM_768_U_VECTOR_BYTES + ML_KEM_768_V_SCALAR_BYTES)
+#  define ML_KEM_1024_CIPHERTEXT_BYTES \
+    (ML_KEM_1024_U_VECTOR_BYTES + ML_KEM_1024_V_SCALAR_BYTES)
+
+/*
+ * Per-variant fixed parameters
+ */
+static const ML_KEM_VINFO vinfo_map[3] = {
+    {
+        "ML-KEM-512",
+        ML_KEM_512_VECTOR_BYTES,
+        ML_KEM_512_PRIVATE_KEY_BYTES,
+        ML_KEM_512_PUBLIC_KEY_BYTES,
+        ML_KEM_512_CIPHERTEXT_BYTES,
+        ML_KEM_512_U_VECTOR_BYTES,
+        sizeof(struct ossl_ml_kem_512_puballoc_st),
+        sizeof(struct ossl_ml_kem_512_prvalloc_st),
+        ML_KEM_512,
+        512,
+        ML_KEM_512_RANK,
+        ML_KEM_512_DU,
+        ML_KEM_512_DV,
+        ML_KEM_512_RNGSEC
+    },
+    {
+        "ML-KEM-768",
+        ML_KEM_768_VECTOR_BYTES,
+        ML_KEM_768_PRIVATE_KEY_BYTES,
+        ML_KEM_768_PUBLIC_KEY_BYTES,
+        ML_KEM_768_CIPHERTEXT_BYTES,
+        ML_KEM_768_U_VECTOR_BYTES,
+        sizeof(struct ossl_ml_kem_768_puballoc_st),
+        sizeof(struct ossl_ml_kem_768_prvalloc_st),
+        ML_KEM_768,
+        768,
+        ML_KEM_768_RANK,
+        ML_KEM_768_DU,
+        ML_KEM_768_DV,
+        ML_KEM_768_RNGSEC
+    },
+    {
+        "ML-KEM-1024",
+        ML_KEM_1024_VECTOR_BYTES,
+        ML_KEM_1024_PRIVATE_KEY_BYTES,
+        ML_KEM_1024_PUBLIC_KEY_BYTES,
+        ML_KEM_1024_CIPHERTEXT_BYTES,
+        ML_KEM_1024_U_VECTOR_BYTES,
+        sizeof(struct ossl_ml_kem_1024_puballoc_st),
+        sizeof(struct ossl_ml_kem_1024_prvalloc_st),
+        ML_KEM_1024,
+        1024,
+        ML_KEM_1024_RANK,
+        ML_KEM_1024_DU,
+        ML_KEM_1024_DV,
+        ML_KEM_1024_RNGSEC
+    }
+};
+
+# define DEGREE ML_KEM_DEGREE
+static const int kPrime = ML_KEM_PRIME;
+
+/*
+ * Remainders modulo `kPrime`, for sufficiently small inputs, are computed in
+ * constant time via Barrett reduction, and a final call to reduce_once(),
+ * which reduces inputs that are at most 2*kPrime and is also constant-time.
+ */
+# define BARRETT_SHIFT (2 * ML_KEM_LOG2PRIME)
+static const unsigned kBarrettShift = BARRETT_SHIFT;
+static const size_t   kBarrettMultiplier = (1 << BARRETT_SHIFT) / ML_KEM_PRIME;
+static const uint16_t kHalfPrime = (ML_KEM_PRIME - 1) / 2;
+static const uint16_t kInverseDegree = ML_KEM_INVERSE_DEGREE;
```

Comment:
All #defines should be at the top of this file after the includes..
The constants should be after that,,

Reply:
Moved these defines.